api-sandbox/Main.elm

```diff
@@ -17,59 +17,76 @@ main =
 
 type Msg
     = GetBlockNumber
-    | Web3Response String
+    | GetBlockNumberResponse String
+    | Web3Response Web3.Response
 
 
 type alias Model =
-    { blockNumber : Maybe Int
-    , web3 : Web3.Model
-    , response : String
+    { blockNumber : Maybe String
+    , web3 : Web3State Msg
+    , error : Maybe String
     }
 
 
 init : ( Model, Cmd Msg )
 init =
     { blockNumber = Nothing
     , web3 = Web3.init
-    , response = "No Blocks yet"
+    , error = Nothing
     }
         ! []
 
 
 view : Model -> Html Msg
 view model =
     div []
-        [ div [] [ viewBlockNumber model ]
+        [ div [] [ text <| viewBlockNumber model ]
         , button [ onClick GetBlockNumber ] [ text "Get Block" ]
-        , div [] [ text model.response ]
+        , viewError model
         , br [] []
         , div [] [ text "Converting a type to look like a function. Perhaps useful." ]
         , div [] [ text <| web3Func <| deCapitalize <| toString GetBlockNumber ]
         ]
 
 
-viewBlockNumber : Model -> Html Msg
+viewBlockNumber : Model -> String
 viewBlockNumber model =
-    case model.blockNumber of
+    model.blockNumber |> Maybe.withDefault "Press button plz"
+
+
+viewError : Model -> Html Msg
+viewError model =
+    case model.error of
         Nothing ->
-            text "Press button plz"
+            span [] []
 
-        Just n ->
-            text (toString n)
+        Just err ->
+            div [] [ text err ]
 
 
 update : Msg -> Model -> ( Model, Cmd Msg )
 update msg model =
     case msg of
         GetBlockNumber ->
             let
-                request =
-                    Web3.Request "eth.getBlockNumber" [] 1
+                ( web3_, cmd ) =
+                    Web3.getBlockNumber model.web3 GetBlockNumberResponse
             in
-                model ! [ web3request request ]
+                { model | web3 = web3_ } ! [ cmd ]
+
+        GetBlockNumberResponse blockNumber ->
+            ( { model | blockNumber = Just blockNumber }, Cmd.none )
+
+        Web3Response { id, result } ->
+            -- TODO
+            -- we know this will be a string at the moment, but will likely need
+            -- custom decoders for each function (some results are objects)
+            case Web3.handleResponse model.web3 id of
+                Nothing ->
+                    { model | error = Just "could get Msg from web3 state" } ! []
 
-        Web3Response response ->
-            { model | response = response } ! []
+                Just msg ->
+                    update (msg result) model
```

Comment:
This was a nice find. It means we can keep the web3 response msg types in the top-level `Msg`